Source/Project64/N64 System/Mips/Pif Ram.cpp

```diff
@@ -297,19 +297,16 @@ void CPifRam::SI_DMA_READ()
 			{
 				continue;
 			}
-			RDRAM[RdramPos ^3] = m_PifRam[count];
+			RDRAM[BES(RdramPos)] = m_PifRam[count];
 		}
 	}
 	else
 	{
-		for (size_t i = 0; i < 64; i += 4)
-		{
-			unsigned __int32 pif_ram_dword;
-			std::memcpy(&pif_ram_dword, &PifRamPos[i], sizeof(unsigned __int32));
-
-			pif_ram_dword = swap32by8(pif_ram_dword);
+		register size_t i;

```

Comment:
You may also have noticed a common fundamental thing between those sources.
They all think more like human beings, less like computers.

Sometimes what they see as more readable, is to small parsers also.  Other times, not really.

>You did not ask why, you asked what.

Wrong, I said:  _unless you can find a reason to suggest that there's anything to lose_

To which you replied by re-stating that C++ deprecated it. That's recursive.

If you had asked me for a reason why there's anything to lose from doing C99's `for (int i = ...` as you stated, and I replied "Aside from it being deprecated?" (assuming some crazy, hypothetical C89-compatible standard chose to deprecate it, whether you personally liked it or not), would you not find that reasoning to be recursive?

*What* is not important.  The question always (and clearly the one I asked) is why.

Should a specification make any wrong or inaccurate turns, then more important is the ability to just not use that spec.  If it maintains backwards compatibility, conforming to it is easier.  Otherwise, how would you react.

>[Unfortunately the ISO C++ Standard Committee does not agree](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#-p3-express-intent)

You mean [this](https://github.com/isocpp/CppCoreGuidelines/blob/master/CppCoreGuidelines.md#Res-recycle)?

Then yes, they do disagree with what I just said.

Unfortunately for them, their "reason" is poorly developed and has little explanation.
I would think that proper "recycling" would look more like this:
```c
int i;

for (i =   0; i <  200; i++) { x[i] = f(i); }
for (i = 400; i <  600; i++) { y[i] = g(i); }
for (i = 800; i < 1000; i++) { z[i] = h(i); }
```
It is fact that debugging this code is easier by simply changing the type of `i`, in this form.

>[nor does CRT (even for C)](https://www.securecoding.cert.org/confluence/display/c/DCL19-C.+Minimize+the+scope+of+variables+and+functions)

I think that declaring a variable at all, anywhere in your code, exposes it to hazards.

Sometimes maintenance hazards, sometimes "security" hazards of form you cite.
Just depends which type of hazard you prefer minimizing the risk of.

Off topic:  Here is another CERT page which I found was somewhat questionable:
https://www.securecoding.cert.org/confluence/display/c/FIO19-C.+Do+not+use+fseek%28%29+and+ftell%28%29+to+compute+the+size+of+a+regular+file

They dictate that the POSIX `fstat()` should be used in place of `fseek()` because of the fact that fseek() to the end of a file is undefined for **non-POSIX** implementations.  c wut they did thar?

>nor does Google outside of object construction where it makes sense, it's also an advisory rule in the MISRA-C guidelines to narrow scope where possible, etc, etc. I could go on and on listing examples.

I am sure you could and that most of them are perfectly credible sources.

But I have not and never will see C programming (cannot say for C++, though) as a question of what other human beings think, but as a question of how I am sure computer logic thinks.  I don't believe for a second that `for (size_t i = ...` is any less ambiguous or open to multiple possibilities of interpretation (or even personal language implementation) than just declaring the variable first.

>If such a practice was more beneficial in the long run, I'm sure your preference would be the one accepted by the standardization committee.

As far as standards committees are concerned, "beneficial" needs extra pragmatism.

They are more interested in the future because they have the power to control it.  You might find this theory is paranoid, but it's really just human nature.  To them, portability means very little if it's only because of backwards compatibility against something they would rather have set their own way.

>If you want the why on that, see [here](http://www.open-std.org/jtc1/sc22/wg21/docs/papers/2014/n4193.html#809) (section 809) and [here](https://isocpp.org/files/papers/n4340.html)

OK, so based on those links, their reasoning is that the rare plausibility of having a unique use for C's version of `register` could, in theory, be replaced by a more concrete and much more useful tool for design/optimization in a future version of the C++ specification.

Certainly possible, but I am glad this "deprecation" hasn't removed the old future yet until they can indeed confirm something to replace it.

Besides, I can take a page from your book and propose that not using the `register` keyword opens up part of your code to accidentally "taking the address" of a variable.  In ANSI C, taking the address of a variable whose declaration contains `register` class is invalid, as you cannot fetch a memory pointer for a CPU register.  (Similarly, you probably use `const` to prevent accidentally writing to a variable you find no plausibility in writing to.)